File: src/Polynomial Multiplication 2/polymul2.cpp
// Using std::complex

#include <bits/stdc++.h>
using namespace std;

vector<complex<double>> fft(vector<complex<double>> coef, bool inverse) { // n is guaranteed a power of 2
    int n = coef.size();
    
    if (n == 1)
        return coef;

    complex<double> z(0,0);
    vector<complex<double>> pe(n/2,z), po(n/2,z);
    for (int i = 0; i < n/2; i++) {
        pe[i] = coef[2*i];
        po[i] = coef[2*i+1];
    }

    vector<complex<double>> ye = fft(pe, inverse), yo = fft(po, inverse), y(n,z); // Recursion
    double angle = (inverse ? -2 : 2)*M_PI/n;
    complex<double> w(cos(angle), sin(angle)), wj = 1;

    for (int j = 0; j < n/2; j++) {
        y[j] = ye[j] + wj*yo[j];
        y[j+n/2] = ye[j] - wj*yo[j];
        wj *= w;
    }

    return y;
}

vector<int> multiply(vector<complex<double>> p1, vector<complex<double>> p2) {
    int n = p1.size(); // which is also p2.size()
    
    vector<complex<double>> fft1 = fft(p1, false), fft2 = fft(p2, false), fftr;

    for (int i = 0; i < n; i++)
        fftr.push_back(fft1[i]*fft2[i]);

    vector<complex<double>> ifftr = fft(fftr, true);

    vector<int> result;
    for (int i = 0; i < n; i++)
        result.push_back(round(real(ifftr[i])/n));

    return result;
}

int main() {
    ios_base::sync_with_stdio(false); 
    cin.tie(NULL);

    int t, d1, d2;
    double c;

    cin >> t;

    while (t--) {
        cin >> d1;
        vector<complex<double>> c1, c2;

        for (int i = 0; i <= d1; i++) {
            cin >> c;
            c1.push_back((complex<double>) c);
        }

        cin >> d2;
        for (int i = 0; i <= d2; i++) {
            cin >> c;
            c2.push_back((complex<double>) c);
        }

        int n = (int) pow(2,ceil(log2(d1+d2+1))); // fill to the nearest power of 2
        for (int i = 0; i < n-d1-1; i++)
            c1.push_back(0);
        
        for (int i = 0; i < n-d2-1; i++)
            c2.push_back(0);

        cout << d1+d2 << endl;
        vector<int> result = multiply(c1,c2);
        for (int i = 0; i <= d1+d2; i++)
            cout << result[i] << " ";
        cout << endl;
    }
    return 0;
}

File: src/Polynomial Multiplication 2/polymul2.py
from math import *

# String representation of polynomial
def rep(p):
    n = len(p)
    sb = []
    for i in range(n):
        if p[i] not in [-1, 1]:
            sb.append(str(p[i]))
        elif p[i] == -1:
            sb.append('-')
        if i > 0:
            sb.append(f'x^{i}')
        if i < n-1 and p[i+1] > 0:
            sb.append('+')
    return ''.join(sb).replace('^1','')

# FFT, handles both FFT and IFFT
def fft(coef, inverse=False):
    n = len(coef)
    if n == 1:
        return coef
    y_e, y_o = fft(coef[::2], inverse), fft(coef[1::2], inverse)
    y = [0]*n
    for i in range(n//2):
        wj = complex(cos((2-4*inverse)*pi*i/n), sin((2-4*inverse)*pi*i/n))
        y[i] = y_e[i] + wj * y_o[i]
        y[i + n//2] = y_e[i] - wj * y_o[i]
    return y

# Multiply two polynomials
# x^2 + 3x + 7 -> [7, 3, 1]
def mult(p1, p2):
    m = len(p1) + len(p2) - 1
    n = 2**(len(bin(m)) - 2)
    p1 = p1 + [0]*(n - len(p1))
    p2 = p2 + [0]*(n - len(p2))
    fft1, fft2 = fft(p1), fft(p2)
    return [t.real/n for t in fft([fft1[i]*fft2[i] for i in range(n)], inverse=True)][:m]
    
q = int(input())
for _ in range(q):
    input()
    p1 = list(map(int, input().split()))
    input()
    p2 = list(map(int, input().split()))
    m = mult(p1, p2)
    print(len(m) - 1)
    print(*list(map(round, m)))